src/executor_cleanup.c

```diff
@@ -0,0 +1,169 @@
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   executor_cleanup.c                                 :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: sqiu <sqiu@student.42vienna.com>           +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2023/08/01 17:11:28 by sqiu              #+#    #+#             */
+/*   Updated: 2023/08/07 16:49:02 by sqiu             ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+/**
+ * @file executor_cleanup.c
+ * @brief Functions to clean up executor module after planned/unplanned abort.
+ */
+
+#include "mod_executor.h"
+
+void	ft_cleanup_cmd(t_cmd *cmd)
+{
+	ft_close(cmd->fd_in);
+	ft_close(cmd->fd_out);
+	ft_plug_pipe(cmd, 1);
+	if (cmd->heredoc)
+		free(cmd->heredoc);
+}
+
+/* This function terminates the program and displays an error message 
+according to the point in the program where the error happened. */
+
+/* void	terminate(char *s)
+{
+	if (errno == 0)
+		ft_putendl_fd(s, 2);
+	else
+		perror(s);
+	exit(1);
+} */
+
+/* This function closes all open file descriptors, frees allocated memory
+space in the heap and unlinks any files at the current point. It then
+gives out an error message and terminates the program. */
+
+/* void	abort_mission(t_meta *meta, char *s)
+{
+	if (meta->here_doc)
+		unlink(".tmp_heredoc");
+	ft_close(meta->fd_in);
+	ft_close(meta->fd_out);
+	free(meta->cmds);
+	terminate(s);
+} */
+
+/* This function frees the cmd_paths and calls 'abort_mission' upon an
+error. */
+
+/* void	no_senor(t_meta *meta, char *s)
+{
+	int	i;
+
+	i = -1;
+	while (meta->cmd_paths[++i])
+		free(meta->cmd_paths[i]);
+	free(meta->cmd_paths);
+	abort_mission(meta, s);
+} */
+
+/* This function frees the arguments in the current command and 
+calls 'no senor' upon an error. */
+
+/* void	mamma_mia(t_meta *meta, char *s)
+{
+	int	i;
+
+	i = -1;
+	while (meta->cmds[meta->i].arg[++i])
+		free(meta->cmds[meta->i].arg[i]);
+	free(meta->cmds[meta->i].arg);
+	no_senor(meta, s);
+} */
+
+/* This function closes remaining file descriptors and frees allocated 
+memory space. */
+
+/* void	cleanup(t_meta *meta)
+{
+	int	i;
+
+	if (meta->here_doc)
+		unlink(".tmp_heredoc");
+	ft_close(meta->fd_in);
+	ft_close(meta->fd_out);
+	i = -1;
+	while (meta->cmd_paths[++i])
+		free(meta->cmd_paths[i]);
+	free(meta->cmd_paths);
+	free(meta->cmds);
+} */
+
+/**
+ * @brief Unlinks the temporary heredoc file created.
+ * 
+ * Returns provided error code.
+ * @param name 		Name of heredoc to be unlinked.
+ * @param err 		Provided error code.
+ * @return t_err 	err
+ */
+t_err	ft_unlink_heredoc(char *name, t_err err)
```

Comment:
After freeing name? Isnt that the case already once it is freed?

Reply:
free() only frees the malloc'd space associated with the value (address) of the passed pointer. Technically free() cannot change the value of the pointer, since you just pass the value and not the reference to the pointer itself.
That's also the reason why we would need to pass a char** to set it to NULL.

If we never look at the pointer again it doesn't make a difference, but since we already have a dedicated cleanup function, we should also set the pointer to NULL.

Reply:
Very good, ran some tests and can confirm that simply freeing the string sets its address to some garbage, not NULL. 

Technically it does not matter whether we set it to NULL though as it will either be overwritten by a new heredoc name or never touched again. But out of principle to have everything neatly reset, just as we set closed fds to -1, the desired performance is now implemented.